sycl/doc/SYCLPluginInterface.md

```diff
@@ -0,0 +1,178 @@
+pointers into a PI interface dispatch table - one per plugin - and uses this
+table to dispatch to the device(s) covered by the corresponding plugin. Further
+sections give complete descriptions of all entry points as well as the dispatch
+table.
+
+PI is based on a subset of OpenCL 1.2 runtime specification, it follows its
+platform, execution and memory models in all aspects except those explicitly
+mentioned in this document. A part of PI API types and functions have exact
+matches in OpenCL. Whenever there is such a match, the semantics also fully
+matches unless the differences are explicitly specified in this document. While
+PI has roots in OpenCL, it does have many differences, and the gap is likely
+to grow, for example in the areas of memory model and management, program
+management.
+
+## Discovery and linkage of PI implementations
+
+![PI implementation discovery](images/SYCL_plugin_discovery.svg)
+
+Device discovery phase enumerates all available devices and their features by
+querying underlying plugins found in the system. This process is only performed
+once before any actual offload is attempted.
+
+### Plugin discovery
+
+Plugins are physically dynamic libraries stored somewhere in the system where
+the SYCL runtime runs. TBD - describe the process in details.
+
+#### Plugin binary interface
+TBD - list and describe all the symbols plugin must export in order to be picked
+up by the SYCL runtime for offload.
+
+#### OpenCL plugin
+
+OpenCL plugin is a usual plugin from SYCL runtime standpoint, but its loading
+and initialization involves a nested discovery process which finds out available
+OpenCL implementations. They can be installed either in the standard Khronos
+ICD-compatible way or not, and the OpenCL plugin can hook up with both.
+
+##### Linking to ICD-compatible OpenCL installations
+
+The plugin dynamically loads the OCL ICD loader - libOpenCL.so from system
+paths, and on success it asks the loader to provide all OpenCL implementations
+installed on the system. To do that it finds and calls the "clGetPlatformIDs"
+in the ICD loader and records the returned platform objects for later use. The
+plugin may choose to dynamically link with libOpenCL.so at build time.
+ 
+##### Linking to ICD-incompatible OpenCL installations
+
+For the OpenCL plugin to hook up such installation, it must be described in a
+configuration file similarly to plugin configuration file. *TBD - location and
+format of custom OpenCL installation config file*. The plugin will dynamically
+load the dynamic library pointed to in the configuration file and query it for
+OpenCL API function addresses similarly ICD loader. In more details, it will:
+- search for  the ```clGetExtensionFunctionAddress``` function in the 
+  library and call it with the string *“clIcdGetPlatformIDsKHR”* to find
+  the ```clIcdGetPlatformIDsKHR``` function pointer
+- then, using ```clIcdGetPlatformIDsKHR```, find the set of
```

Comment:
From the text above, seems to refer to an OpenCL implementation that has the `cl_khr_icd` extension but is not registered in `/etc/OpenCL/vendors` or wherever this happens on Windows. 
Seems to me that, for it to be really flexible, it shouldn't rely on the cl_khr_icd extension, and simply work with any libOpenCL.so that exposes the `clGetPlatforms` function.

Reply:
ICD-incompatible runtimes are those like packaged with SYCL distribution but not installed in such a way that they can be found by the standard ICD loader.

> Seems to me that, for it to be really flexible, it shouldn't rely on the cl_khr_icd extension, and simply work with any libOpenCL.so that exposes the clGetPlatforms function.

hope we can we change this later